### PR TITLE
Plugins: Add apiserver info to settings when it exists

### PR DESCRIPTION
**Pull request by ryantxu**
This replaces https://github.com/grafana/grafana/pull/87639

This updates the settings model so it is more clearly a indication of where the apiserver is running with which versions:
image

### CODE REVIEW

File: pkg/api/dtos/plugins.go [hunk 1]
```diff
@@ -28,7 +28,7 @@ type PluginSetting struct {
 	SignatureType   plugins.SignatureType   `json:"signatureType"`
 	SignatureOrg    string                  `json:"signatureOrg"`
 	AngularDetected bool                    `json:"angularDetected"`
-	APIVersion      string                  `json:"apiVersion"`
```
Comment (ryantxu) [inline pkg/api/dtos/plugins.go:31]: the backend is not a single apiVersion -- it may have a few and a preferred version

**Comment by ryantxu** [comment]
@andresmgot -- See https://github.com/grafana/grafana/pull/87639#discussion_r1606484446 -- that thread got lost when I made a new PR here.

File: pkg/api/dtos/plugins.go (hunk 1 above)
Comment (andresmgot) [reply pkg/api/dtos/plugins.go:31]: why not simplify this just using the `preferredVersion` here? In which case the frontend would be interested in knowing all the versions the backend supports?

Also, why duplicating this information? This is currently available in the endpoing `/apis` for the datasource backend endpoint.

File: pkg/apimachinery/apis/common/v0alpha1/resource.go [hunk 2]
```diff
@@ -26,10 +26,13 @@ func NewResourceInfo(group, version, resourceName, singularName, kind string,
...
 
-func (info *ResourceInfo) WithGroupAndShortName(group string, shortName string) ResourceInfo {
+func (info *ResourceInfo) WithGroupAndShortName(group, version, shortName string) ResourceInfo {
+	if version == "" {
+		version = info.version
+	}
 	return ResourceInfo{
 		group:        group,
-		version:      info.version,
+		version:      version,
```
Comment (andresmgot) [inline pkg/apimachinery/apis/common/v0alpha1/resource.go:35]: I am not sure I understand why this change

File: pkg/registry/apis/datasource/register.go [hunk 3]
```diff
@@ -224,7 +227,9 @@ func (b *DataSourceAPIBuilder) getPluginContext(ctx context.Context, uid string)
 	if err != nil {
 		return backend.PluginContext{}, err
 	}
-	return b.contextProvider.PluginContextForDataSource(ctx, instance)
+	pctx, err := b.contextProvider.PluginContextForDataSource(ctx, instance)
+	pctx.APIVersion = b.connectionResourceInfo.APIVersion()
```
Comment (andresmgot) [inline pkg/registry/apis/datasource/register.go:231]: This can probably be in a different PR? It's unrelated to having apiserver info in the plugin DTO.

File: pkg/api/dtos/plugins.go (hunk 1 above)
Comment (andresmgot) [reply pkg/api/dtos/plugins.go:31]: replied already here: https://github.com/grafana/grafana/pull/87639#discussion_r1606484446

Not sure if the frontend needs to handle this complexity (multiple api versions) but it's syntactically correct so if we want to move forward with this it can be okay

Comment (ryantxu) [reply pkg/api/dtos/plugins.go:31]: Alternatively we can keep this out of the plugins DTO entirely -- I want to make sure we do not read it to mean more than it does.  The issue I have with the simplified version is that it encourages thinking about frontend+backend as one deployment rather than two things: client and server

File: pkg/apimachinery/apis/common/v0alpha1/resource.go (hunk 2 above)
Comment (ryantxu) [reply pkg/apimachinery/apis/common/v0alpha1/resource.go:35]: used here:  https://github.com/grafana/grafana/pull/88041/files#diff-4f8682b7549ec2d252d819509508d94153e4e75210d627b625ff95800af6f2dfR169

it sets the version rather than locked at v0alpha1

File: pkg/registry/apis/datasource/register.go (hunk 3 above)
Comment (ryantxu) [reply pkg/registry/apis/datasource/register.go:231]: 👍🏻 -- i can split it out soon, already a few too many open PRs for hte same topic 🤣

File: pkg/api/dtos/plugins.go (hunk 1 above)
Comment (andresmgot) [reply pkg/api/dtos/plugins.go:31]: we need to have _something_ here if we want to set the value in queries / configs. The frontend will use the `preferredVersion` anyway and use it. I don't see a use case for the fronted to use the rest of the information so why adding it?

Comment (marefr) [reply pkg/api/dtos/plugins.go:31]: What do we need to move this forward? Agreement on either only including preferred version or plus api group and versions? 

Given this is new/experimental can we start with just adding what we actually need right now and iterate figuring out the rest? As long as we don't officially support any new api response field we can add/remove/change how many times we want/need.

Comment (andresmgot) [reply pkg/api/dtos/plugins.go:31]: As far as I understand from this comment: https://github.com/grafana/grafana/pull/87961#issuecomment-2124576258 we may want to just remove the `apiVersion` from the `/api/*` endpoints and leave it for the new ones under `/apis/*` :shrug:

**Comment by diegommm** [comment]
Hi! Greetings from the Release Guild. Just letting you know I'm updating your PR's milestone to [11.2.x](https://github.com/grafana/grafana/milestone/529). If you need to see your changes in a previous version, please consider adding a backport label.

Thank you!

**Comment by ryantxu** [comment]
@andresmgot -- for the query migration workflow, I think this makes sense again.

> In which case the frontend would be interested in knowing all the versions the backend supports?

The frontend knows how to support some version(s), and it will need to make sure the backend supports the same thing.  This is not relevant when the frontend+backend are always deployed together, but we should stop assuming that and work as if the backend was a service and the frontend a client.

Duplicated in /apis/? -- true... maybe we should just attach the group, then the client looks up versions from /apis/{group} -- either approach makes sense to me

**Comment by ryantxu** [comment]
See https://github.com/grafana/grafana/pull/89772
